[WebProfiler] added support for window.fetch calls in ajax section

ivoba opened the pull request:
| Q | A |
| --- | --- |
| Branch? | "master" |
| Bug fix? | yes |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #17444 |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

This adds support for window.fetch calls to the Ajax section of the WebProfiler toolbar.

Credits to @tbopec for  implementation :)

ivoba committed:
added support for window.fetch calls in ajax section

fabpot commented:
@javiereguiluz Can you validate this one?

fabpot commented:
@robfrawley Can you test it?

nicolas-grekas commented:
Could anyone please test this one?

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -236,6 +236,31 @@
+                            stackElement.error = r.status < 200 || r.status >= 400;
+                            stackElement.statusCode = r.status;
+                            stackElement.profile = r.headers.get('x-debug-token');
+                            stackElement.profilerUrl = r.headers.get('x-debug-token-link');
```
Is this compatible with CORS restriction or will it log a security warning ?

fabpot replied:
@romainneutron any ideas?

romainneutron replied:
I'm gonna try this PR

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -236,6 +236,31 @@
 
         {% if excluded_ajax_paths is defined %}
             if (window.XMLHttpRequest && XMLHttpRequest.prototype.addEventListener) {
+                if (window.fetch) {
```
this does not need to be inside the check for ``window.XMLHttpRequest``

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -236,6 +236,31 @@
+                        var stackElement = {
+                            loading: true,
+                            error: false,
+                            url: arguments[0],
```
you need to match the URL against the excluded ajax paths before adding it in the request stack

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -236,6 +236,31 @@
+                            start: new Date()
+                        };
+                        requestStack.push(stackElement);
+                        promise.then(function (r) {
```
what if the promise gets rejected ? The stack element should not stay in loading state in this case

stof commented:
Status: needs work

ivoba committed:
applied suggested fixes:
added error handling on promise catch,
moved fetch block out of XHR block,
fixed method detect,
matched url against excluded urls

ivoba commented:
Applied fixes for @stof `s comments.

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -1,4 +1,4 @@
-<script{% if csp_script_nonce is defined and csp_script_nonce %} nonce={{ csp_script_nonce }}{% endif %}>/*<![CDATA[*/
```
This remove must be reverted. You are breaking the CSP compatibility

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -22,204 +22,204 @@
-            collectionToArray = function (collection) {
-                var length = collection.length || 0,
-                    results = new Array(length);
+                collectionToArray = function (collection) {
```
please revert all these indentation changes

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,77 +235,94 @@
+        if (window.fetch) {
+            var oldFetch = window.fetch;
+            window.fetch = function () {
+                if (!arguments[0].match(new RegExp({{ excluded_ajax_paths|json_encode|raw }}))) {
```
applying the original fetch (and returning the promise) must still be made even when the URL is excluded from the profiling, otherwise you are breaking ``window.fetch``

stof commented:
I will review again when the indentation changes are reverted. They make it very hard to see what actually changed (and if I ignored the whitespace changes, I cannot comment on the diff anymore, which is not convenient)

ivoba commented:
oops how did that happen?
fix is coming...

ivoba commented:
@stof done, sorryy

tsm91 commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,46 @@
+                        var promise = oldFetch.apply(null, arguments);
+
+                        var method = 'GET';
+                        if(arguments[1].method !== undefined){
```
@ivoba The `init` config object is optional (`Promise<Response> fetch(input[, init]);`) therefore this can throw an error. First check if the 2nd argument even exists `arguments[1] && arguments[1].method !== undefined` .

ivoba replied:
@tsm91 correct. i fixed this.

fabpot commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,47 @@
         }
 
         {% if excluded_ajax_paths is defined %}
+
```
extra blank line here

fabpot commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,47 @@
+                var oldFetch = window.fetch;
+                window.fetch = function () {
+                    if (!arguments[0].match(new RegExp({{ excluded_ajax_paths|json_encode|raw }}))) {
+
```
extra blank line here

fabpot commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,47 @@
+
+                        var method = 'GET';
+
+                        if(arguments[1] && arguments[1].method !== undefined){
```
missing space after `if` and before `{`

Gadgetdude commented:
Really looking forward to this fix/feature. Is there any way I can assist?

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,45 @@
+                var oldFetch = window.fetch;
+                window.fetch = function () {
+                    if (!arguments[0].match(new RegExp({{ excluded_ajax_paths|json_encode|raw }}))) {
+                        var promise = oldFetch.apply(null, arguments);
```
this call must be done outside the URL profiling blacklist check (and the return at the end too). Otherwise, you break fetch entirely for URLs blacklisted by the AJAX profiling

ivoba replied:
@stof fixed

romainneutron commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,44 @@
+            if (window.fetch) {
+                var oldFetch = window.fetch;
+                window.fetch = function () {
+                    var promise = oldFetch.apply(null, arguments);
```
`oldFetch.apply(window, arguments);`

romainneutron replied:
Works perfectly, no security error, even in `cors` mode

stof replied:
even better: ``.apply(this, arguments``, keeping the same context

ivoba replied:
fixed using @stof 's approach

romainneutron commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,44 @@
+                            stackElement.profile = r.headers.get('x-debug-token');
+                            stackElement.profilerUrl = r.headers.get('x-debug-token-link');
+                            Sfjs.renderAjaxRequests();
+                        }).catch(function(err) {
```
it's better to use the second argument of `then` instead of using `catch`: if an error is throws by `Sfjs.renderAjaxRequests`, the request would be tagged as erroneous whereas the error came from the code executed after that.

ivoba replied:
fixed

romainneutron commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,44 @@
+                        }).catch(function(err) {
+                            stackElement.loading = false;
+                            stackElement.error = true;
+                        });
```
what about adding the error content as a property and display this error in the toolbar?

ivoba replied:
its a good idea, but maybe we can put this in another PR since its also not implemented for XHR, afaik, and it would need some more adjustments in how we render the table.

romainneutron commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,44 @@
+                            Sfjs.renderAjaxRequests();
+                        }).catch(function(err) {
+                            stackElement.loading = false;
+                            stackElement.error = true;
```
what about adding a `type` so we could make distinction between `ajax (xhr)` and `fetch` in the toolbar. As they do not behave the same way, it may be useful, WDYT?

ivoba replied:
yeah was thinking about that too, but was bit afraid of the extra effort. i see what i can do.

stof replied:
well, the extra work is mainly about figuring out the display. Adding the type property is not hard (it should just be done in both place initializing a stackElement)

ivoba replied:
i added type to stackElement and to display

romainneutron commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,44 @@
+
+            if (window.fetch) {
+                var oldFetch = window.fetch;
+                window.fetch = function () {
```
what about using `fetch` signature (`fetch(url, options):promise`) instead of relying on arguments? I think it's easier to read

ivoba replied:
absolutly, fixed

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/base_js.html.twig:
```diff
@@ -235,6 +235,44 @@
 
         {% if excluded_ajax_paths is defined %}
+
+            if (window.fetch) {
```
should we try to detect native fetch here ? If someone uses a fetch polyfill based on XHR (e.g. the polyfill maintained by github), we would register the request twice otherwise (at the fetch level and at the xhr level)

ivoba replied:
how would you detect this? just check if the browser supports fetch? this needs to happen before any ployfill is loaded, im not sure how to do this.
also as the profiler is a dev tool, how many people develop on outdated browsers?

ivoba replied:
ok i see, i added a check on polyfill `if (window.fetch && window.fetch.polyfill === undefined)`

HeahDude commented:
@ivoba Do you need help with the templates so you can finish this feature for 3.2? :)

ivoba committed:
added type to profiler table, use 2nd argument instead of catch, using fetch signature instead of arguments

ivoba committed:
fixed polyfill detect

stof commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar.css.twig:
```diff
@@ -334,7 +334,7 @@
 }
 .sf-ajax-request-url {
-    max-width: 300px;
+    max-width: 250px;
```
why changing this ?

fabpot commented:
Thank you @ivoba.
